com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+using MLAPI.Serialization.Pooled;
+using MLAPI.Serialization;
+using MLAPI.Configuration;
+using MLAPI.Profiling;
+using System.Collections.Generic;
+using System.Linq;
+using System;
+using System.IO;
+
+namespace MLAPI
+{
+    class MessageBatcher
+    {
+        public class SendStream
+        {
+            public FrameQueueItem Item;
+            public PooledBitStream Stream = new PooledBitStream();
+        }
+
+        // Stores the stream of batched RPC to send to each client, by ClientId
+        private Dictionary<ulong, SendStream> SendDict = new Dictionary<ulong, SendStream>();
+        private PooledBitWriter Writer = new PooledBitWriter(new PooledBitStream());
+
+        // Used to store targets, internally
+        private List<ulong> TargetList = new List<ulong>();
+
+        public void PushLength(int length, ref PooledBitWriter writer)
+        {
+            // If length is single byte we write it
+            if (length < 256)
+            {
+                writer.WriteByte((byte)length); // write the amounts of bytes that are coming up
+            }
+            else
+            {
+                // otherwise we write a 0, and a two-byte length
+                writer.WriteByte(0); // mark larger size
+                writer.WriteByte((byte)(length % 256)); // write the length modulo 256
+                writer.WriteByte((byte)(length / 256)); // write the length divided by 256
+            }
+        }
+
+        public int PopLength(in BitStream messageStream)
+        {
+            byte read = (byte)messageStream.ReadByte();
+            // if we read a non-zero value, we have a single byte length
+            if (read != 0)
+            {
+                return read;
+            }
+            // otherwise, a two-byte length follows
+            int length = messageStream.ReadByte();
```

Comment:
Ugh, we should fix BitStream::ReadByte to return, well, a byte not an int with -1 encoded as fail.